### DISCUSSION TITLE
💡 Shiki Editor

### DISCUSSION THREAD

**Discussion by magic-akari** [shikijs/shiki#650] [Show and tell]
Hello, I am amazed by the powerful extensibility of Shiki.
I have created a lightweight code editor using Shiki and a textarea.

This editor utilizes a textarea for input, leverages Shiki for code highlighting, and incorporates functionalities such as tab key indentation and auto insertion of matching brackets.

You can explore the playground of this editor by visiting the following link: [Shiki Editor Playground](https://shiki-editor.vercel.app)

I am planning to release this editor as an npm package, but I am unsure if there is interest in it, so I am seeking feedback.

Here is the repository source code: https://github.com/magic-akari/shiki-editor

Moreover, I am interested to inquire whether the Shiki team would be open to including this editor within the Shiki organization.

**Comment by antfu** [comment]
That looks nice, but unfortunately, I don't think we could make this official at this moment:

- The repo you linked seems to be at quite an early stage, I think it would be better if it can have some docs to describe how to use it etc.
- "and incorporates functionalities such as tab key indentation and auto insertion of matching brackets" - there are nice - but I am concerned about the scope in the future, knowing that creating a code editor is super challenging (even tho we might not going to create another Codemirror or Monaco, we still need to figure out the line what feature request we should take)

I am happy to link to your project in our docs if you could write some readme guides for it - later, if it's verified by the community that's a common need; we could then discuss again about whether to move to the org. Thank you.

**Reply by magic-akari** [reply]
This is indeed a discussion about the future. 
The project, formerly named with Shiki Editor and now rebranded as Shiki Code, is still in its early stages and undergoing active development.

I am striving to shape ShikiCode into a core editor library that is both simple at its foundation and highly extensible.
It's challenging to draw a definitive line and declare, "We won't support this feature." Instead, I aim to ensure that the core of Shiki Code remains minimal, with any additional functionalities existing in the form of plugins.

Currently, the core functionality of Shiki Code includes line number rendering and scroll synchronization, with the latter being more of a necessary behavior than a feature.
Features like tab key indentation, auto closing, and others have been relocated to subpath under `plugins`. And user-requested features like the <kbd>cmd</kbd> + <kbd>/</kbd> commenting functionality will be implemented as plugins. 

I will listen to user feedback, but I also reserve the right to decline its implementation (∠・ω< )

Furthermore, rather than explicitly stating which features we will support or decline, I believe it is more crucial to establish a clear positioning for Shiki Code. I will document which scenarios are suitable for Shiki Code, while other scenarios may be better suited for Monaco or Codemirror.

My current concern is that ShikiCode exists as a standalone repository. Therefore, I am questioning the appropriateness of using the Shiki name and Shiki logo. While Shiki Code does not have a designed logo at the moment, I envision creating one in the future that resonates with Shiki's design aesthetics.

**Comment by NWYLZW** [comment]
Similarly, I would like to thank [antfu](https://github.com/antfu) for your work on Shiki, and also thank [magic-akari](https://github.com/magic-akari) for inspiring me in this matter. Over the past half month, I have done some minor work to systematize the ideas of shiki-editor and have been developing it as a project for some time. Now, I can provide a small core library for users to use, while also offering enhanced editing functionality as plugins, as mentioned by magic-akari above. Users can selectively use the parts they need.
I have now released the first [pre-release](https://github.com/NWYLZW/shikitor/releases/tag/0.0.3) of this project and provided some documentation on how to use it in the [README.md](https://github.com/NWYLZW/shikitor?tab=readme-ov-file#shikitor). If anyone is interested, you can find it here. This project may be a long-term challenge for me, so I have planned some tasks to complete it. If anyone is interested in this project, feel free to initiate a PR to contribute.

<img width="1728" alt="image" src="https://github.com/shikijs/shiki/assets/51358815/23c4ec58-5458-4079-828c-34a93d44a73f">

**Comment by kinthaiofficial** [comment]
Shiki as an editor's syntax highlighter is an exciting direction — it would bring the same grammar quality that powers VS Code to the web editor space.

The key design challenge is performance mode: Shiki uses TextMate grammars which are synchronous and CPU-bound. For an editor that needs to re-highlight on every keystroke, this needs careful handling.

**Current approach for editor integration**:

```typescript
import { createHighlighter } from 'shiki';

// Initialize once (expensive — loads language grammars)
const highlighter = await createHighlighter({
  themes: ['github-dark'],
  langs: ['typescript', 'python', 'rust'],
});

// Highlight is synchronous after init (fast for incremental)
function highlight(code: string, lang: string): string {
  return highlighter.codeToHtml(code, { lang, theme: 'github-dark' });
}
```

**For editor use (incremental highlighting)**:
```typescript
// The challenge: full re-highlight on every edit is too slow for large files
// Need incremental: only re-highlight changed lines + context

// Approach 1: Worker thread
const worker = new Worker('/shiki-worker.js');
worker.postMessage({ code, lang });
worker.onmessage = (e) => applyHighlights(e.data);

// Approach 2: Shiki's onigurumaRegistry for incremental tokenization
import { createHighlighterCore } from 'shiki/core';
const h = await createHighlighterCore({ ... });
const tokens = h.codeToTokens(code, { lang });
// Process tokens incrementally
```

**What would make Shiki Editor compelling**:
- Monaco-style multi-language support (detect language automatically)
- TextMate scope → editor theme color mapping
- Semantic token support (LSP semantic highlighting layered over grammar)

Is the goal a standalone editor component, or a CodeMirror/ProseMirror plugin?
